### DISCUSSION TITLE
Is there a way to be generic over the parameters of a generic type?

### DISCUSSION THREAD

**Discussion by fjarri** [python/typing#1634] [Q&A] [answered]
For example, I want a method to accept a callable that takes a `Union` type with any parameters and returns a value of that `Union` type (or, in other words, a value of any type from the union). That is, something like
```
Callable[Type[Union[*P]], Union[*P]]
```
For example, if it is given an argument `Union[int, str]` (the type, not a value of this type), it will return a value of type `int` or of type `str`. 

`ParamSpec` and `TypeVar` doesn't seem to work here. Is it currently possible in Python?

A practical application here is a deserializing hook: it is given a type (possibly a generic, like `Union` of something), and a value of type `Any`, and it tries to deserialize it into that type.

**Comment by TeamSpen210** [comment, marked as answer]
Unfortunately no. First, something like `Union[...]` isn't valid to pass to a function accepting `type` - that only accepts actual classes. The solution to that is likely a new typing construct - [`TypeForm`](https://github.com/python/mypy/issues/9773). The PEP for that is still in draft stage though. Secondly `Union[*Ts]` isn't supported itself. This was [mentioned](https://github.com/python/mypy/issues/16720) in early versions of the TypeVarTuple PEP, but was removed due to the limited number of use cases and implementation complexity required.
